### PR TITLE
search: add tensor core and hand-coded opts to beam search

### PR DESCRIPTION
**Pull request by flammit**
This PR depends on #2097.

It also simplifies the search action space to only have GROUP (which implies top) and LOCAL (which implies last).  There's no measurable difference for the matvec ops to use GROUP vs. GROUPTOP and the tensor core structure is now written to only use LASTLOCAL.

### CODE REVIEW

**Comment by flammit** [comment]
Using beam search on a Metal we can get cifar down to 47-48ms on an M2 Max:

```
955   47.35 ms run,    7.92 ms python,   39.42 ms CL,  496.25 loss, 0.000019 LR, 2.17 GB used,   9891.00 GFLOPS
956   47.30 ms run,    7.08 ms python,   40.23 ms CL,  495.75 loss, 0.000014 LR, 2.17 GB used,   9900.40 GFLOPS
957   47.12 ms run,    6.84 ms python,   40.28 ms CL,  503.50 loss, 0.000009 LR, 2.17 GB used,   9939.55 GFLOPS
958   47.40 ms run,    7.18 ms python,   40.23 ms CL,  505.50 loss, 0.000005 LR, 2.17 GB used,   9879.75 GFLOPS
959   48.88 ms run,    6.95 ms python,   41.93 ms CL,  502.50 loss, 0.000000 LR, 2.17 GB used,   9582.01 GFLOPS
...
eval     9585/10240 93.60%,    0.40 val_loss STEP=960 (in 22415.96 ms)
eval ema 9596/10240 93.71%,    0.40 val_loss STEP=960
```

File: tinygrad/codegen/linearizer.py [hunk 1]
```diff
@@ -223,82 +229,75 @@ def end_loop(xx:List[Variable]):
...
       # define indexes
       reduce_idxs = [Variable(f"ridx{i}", 0, self.full_shape[i]-1) for i in range(self.first_reduce+len(self.group_for_reduce), self.shape_len-self.upcasted)]
       fake_reduce_idxs = [x*0 for x in reduce_idxs]
 
       # define accumulator
       acc = self.global_load(0, global_idxs+local_idxs+fake_reduce_idxs+upcast_idxs, {ReduceOps.SUM: 0.0, ReduceOps.MAX: -math.inf}[cast(ReduceOps, self.reduceop.op)])
+      val_idxs = global_idxs+local_idxs+fake_reduce_idxs+upcast_idxs
+
+      if self.tensor_core:
+        def calc_tc_idxs(local_size: int, aliases: List[List[int]]):
```
Comment (roelofvandijk) [inline tinygrad/codegen/linearizer.py:248]: Does it make sense to define this elsewhere and cache it (with aliases as tuple of tuples)?

File: tinygrad/codegen/linearizer.py [hunk 2]
```diff
@@ -223,82 +229,75 @@ def end_loop(xx:List[Variable]):
...
-        idxs = global_idxs+local_idxs+reduce_idxs+(this_upcast_idxs[::-1] if self.reverse_upcast_dir else this_upcast_idxs)
-        idxs = [idx*0 if s == 0 else idx for idx,s in zip(idxs,strides)]
-        if DEBUG >= 3: print(f"{localbuf_idx} alias {i}:", idxs)
-        ll = self.global_load(i, idxs)
-        locals_to_store.append((localbuf_idx, idxs, ll))
+        buf_idxs = [idx*0 if s == 0 else idx for idx,s in zip(global_idxs+local_idxs+reduce_idxs+full_upcast_idxs,self.sts[i].real_strides())]
+        if self.tensor_core:
+          min_alias_idx = min(self.local_alias.keys())
+          replace_input_idxs = calc_tc_idxs(self.tensor_core.thread_local_sizes[i-min_alias_idx], self.tensor_core.thread_local_aliases[i-min_alias_idx])
+          for n in range(len(self.tensor_core.threads)):
```
Comment (roelofvandijk) [inline tinygrad/codegen/linearizer.py:281]: This could be a function (duplicate of line 261ff).

File: tinygrad/codegen/optimizer.py [hunk 3]
```diff
@@ -159,87 +159,30 @@ def alias_buffer(self, i, pattern):
...
-      buf1_strides = self.sts[buf1].real_strides()
-      axis_buf0 = [(i,self.full_shape[i],buf1_strides[i]) for i,s in enumerate(buf0_strides) if s == 0 and self.full_shape[i]%8 == 0 and i < self.first_reduce]
-      axis_buf1 = [(i,self.full_shape[i],buf0_strides[i]) for i,s in enumerate(buf1_strides) if s == 0 and self.full_shape[i]%8 == 0 and i < self.first_reduce]
-      if axis_buf0 and axis_buf1 and self.full_shape[self.first_reduce]%8 == 0 and self.shape_len-self.first_reduce == 1:
-        if DEBUG >= 3: print("METAL TENSOR CORES", axis_buf0, axis_buf1)
+        if has_cast and not(isinstance(self.reduceop.src[0], LazyOp) and self.reduceop.src[0].op == UnaryOps.CAST and self.reduceop.src[0].arg[0] == tc.dtype_out): continue
+        mul_op = self.reduceop.src[0].src[0] if has_cast else self.reduceop.src[0]
+
+        if not(isinstance(mul_op, LazyOp) and mul_op.op == BinaryOps.MUL): continue
+        if not(isinstance(mul_op.src[0], LazyOp) and mul_op.src[0].op == BufferOps.MEM and mul_op.src[0].arg.dtype == tc.dtype_in): continue
```
Comment (roelofvandijk) [inline tinygrad/codegen/optimizer.py:172]:
```
for i in [0,1]:
  if not(isinstance(mul_op.src[i], LazyOp) and mul_op.src[i].op == BufferOps.MEM and mul_op.src[i].arg.dtype == tc.dtype_in): continue
```

Maybe even:
```
for src in mul_op.src[:2]:
  if not(isinstance(src, LazyOp) and src.op == BufferOps.MEM and src.arg.dtype == tc.dtype_in): continue
```

File: tinygrad/codegen/optimizer.py [hunk 4]
```diff
@@ -159,87 +159,30 @@ def alias_buffer(self, i, pattern):
...
+        if not(isinstance(mul_op.src[1], LazyOp) and mul_op.src[1].op == BufferOps.MEM and mul_op.src[1].arg.dtype == tc.dtype_in): continue
+        buf0, buf1 = self.bufs.index(cast(LazyOp, mul_op.src[0].arg)), self.bufs.index(cast(LazyOp, mul_op.src[1].arg))
+        buf0_strides, buf1_strides = self.sts[buf0].real_strides(), self.sts[buf1].real_strides()
+        axis_buf0 = [(i,self.full_shape[i],buf1_strides[i]) for i,s in enumerate(buf0_strides[:self.first_reduce]) if s == 0 and self.full_shape[i]%tc.dims[0] == 0]
```
Comment (roelofvandijk) [inline tinygrad/codegen/optimizer.py:176]: `s == self.full_shape[i]%tc.dims[0] == 0`

File: tinygrad/features/search.py [hunk 5]
```diff
@@ -8,15 +8,15 @@
...
+actions += flatten([[Opt(op=OptOps.GROUP, axis=axis, amt=amt) for amt in [16,256]] for axis in range(3)])
+actions += [Opt(op=OptOps.GROUP, axis=0, amt=amt) for amt in [4,8]]
 actions += [
   Opt(op=OptOps.LOCAL, axis=0, amt=32),
-  Opt(op=OptOps.GROUP, axis=0, amt=4), Opt(op=OptOps.GROUP, axis=0, amt=8), Opt(op=OptOps.GROUP, axis=1, amt=8),
-  Opt(op=OptOps.GROUPTOP, axis=0, amt=16), Opt(op=OptOps.GROUPTOP, axis=0, amt=256),
-  Opt(op=OptOps.GROUPTOP, axis=1, amt=16), Opt(op=OptOps.GROUPTOP, axis=1, amt=256),
-  Opt(op=OptOps.GROUPTOP, axis=2, amt=16), Opt(op=OptOps.GROUPTOP, axis=2, amt=256),
   Opt(op=OptOps.UPCASTMID, axis=1, amt=4),
+  Opt(op=OptOps.TC, axis=0, amt=0),
```
Comment (geohot) [inline tinygrad/features/search.py:18]: So this isn't how this should be done, this is hiding complexity.

The individual ops of tensor core should be in the search space.

**Comment by geohot** [comment]
If GROUP/GROUPTOP can be unified, this should be in a separate PR.

**Comment by flammit** [comment]
> If GROUP/GROUPTOP can be unified, this should be in a separate PR.

Yup - I was gonna split out the GROUP/GROUPTOP along with LOCAL simplification too as pre-requisite PR to this one.

**Comment by flammit** [comment]
Closing this out and will start with the search space simplifications first.

**Pull request closed without merging**